### DISCUSSION TITLE
UpdateTypeManager was involved when updateType is not available

### DISCUSSION THREAD

**Discussion by lyonlai** [renovatebot/renovate#24684] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:36.107.1-slim

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

We have a monorepo making progress rollout for packages inside. And we found that packages in group with `matchPackagePrefixes` and `matchUpdateTypes` was considered disabled before fetching update for packages.

The configuration looks like 

```json5
{ 
  packageRules: [
    { // this rule is mainly for being able to do rollout in chunks for large monorepo
      matchManagers: [
        'gradle'
      ],
      enabled: false
    },
   {
      "groupName": "some group name",
      matchManagers: [
        'gradle'
      ],
      matchPackagePrefixes: [
        'packageA',
        'packageB',
        'packageC',
      ],
      matchUpdateTypes: [
        'minor',
        'patch',
      ],
      enabled: true
    },
  ]
}
```

With the current logic, `packageA`, `packageB` and `packageC` result in `Dependency {{packageName}} is disabled` just after dependencies have been extracted and right before [fetchDepUpdates](https://github.com/renovatebot/renovate/blob/848a0b03763f79eb3546a4993cd78e8994d00577/lib/workers/repository/process/fetch.ts#L66)  happens. This causes Renovate to close the PR that's originally opened.

tracing down to why this happens leads me to the [`UpdateTypesMatcher`](https://github.com/renovatebot/renovate/blob/848a0b03763f79eb3546a4993cd78e8994d00577/lib/util/package-rules/update-types.ts#L10).  `updateType` is `undefined` at that particular point in time. And the matcher probably should not be involved at that time. Hence in the [original PR](https://github.com/renovatebot/renovate/pull/24683) I made the matcher returns null when the given updateType is undefined. 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
First up: this is intended behavior and if we were to change it, it would change behavior for thousands of users.

Also, this change of semantic would then contradict how other matches work.

I understand how your two rules can lead to confusion, but ultimately your first rule is too greedy. If you add a `matchUpdateTypes` matcher to it then it will delay its evaluation and the two together will behave as you need

**Reply by lyonlai** [reply]
I've tried it out briefly in a test repository and it does seem to make renovate behave as we wanted. I seem to be able to understand the delay mechanism you mentioned, please correct me if I'm wrong. By adding `matchUpdateType` into the first rule, the first rule wouldn't even match and leave the `enabled` option as the default value, which is `true`. Then the package wouldn't be disabled before [fetchDepUpdates](https://github.com/renovatebot/renovate/blob/848a0b03763f79eb3546a4993cd78e8994d00577/lib/workers/repository/process/fetch.ts#L66).  But the rule will be applied once updateType is available update information later down the road.

Though it make renovate behaves as wanted, it still feeling a bit odd for me:

1.  This solution relies on `enabled` option's default value. Other option combinations might still bring similar issues back in the future.  
2.  `matchUpdateTypes` has a list of `allowedValues`, which is all the possible update types from all managers. We will either need to know all the possible values for the manager we are applying, or just fill in all the `allowedValues`. Not to mention that the values of  `allowedValues` grows with Renovate. 
3.  By delaying the evaluation of the first rule, Renovate will go ahead and fetch the update information even for those intended to be disabled packages. It might not be obvious for other projects. But internally we have both complex monorepos and clunky maven repository.  From our metrics the fetching and parsing of those information are not cheap. Would still prefer not fetching the update information for those disabled packages if possible.

**Reply by rarkins** [reply]
Your understanding in the first paragraph is correct.

This problem partly comes about because users may have a different intention for what enabled=false means. For some, they mean they want to disable the dependency and not have it looked up. While for others, it might mean they simply don't want PRs created.

BTW if you had chosen dependencyDashboardApproval=true instead of disabling, then you would not have had the problem that way either. We could maybe document/recommend this approach.

Also I think that supporting a special `["*"]` option for most of these string matchX matchers would make the usability better.

**Reply by lyonlai** [reply]
> This problem partly comes about because users may have a different intention for what enabled=false means. For some, they mean they want to disable the dependency and not have it looked up. While for others, it might mean they simply don't want PRs created.

Would the team consider making another option to be able to control different intents in the future?

In regards of the `dependencyDashboardApproval` that is new to me, with a quick look, seems like it is considered at `processBranch` which has already passed the look up stage. It's good to know but wouldn't make a big difference with adding the updateType. But the dashboard might look busier afterward.

supporting `['*']` would be nice and def will cover the growing `updateType` problem.

So one thing still confuses me is that, why would the `UpdateTypesMatcher` get involved when the input value `updateType` is not presented? Would you mind show me the rational behind?

**Reply by rarkins** [reply]
Your rule says `matchUpdateTypes=foo` and it fails to match (no updateType is present) so it returns false. This is the same across every type of rule and is not unique to matchUpateTypes

**Reply by lyonlai** [reply]
I see. Thank you.
